### PR TITLE
Enable Correlations for Groups

### PR DESCRIPTION
**Pull request by neilkakkar**
## Changes

~~TODO: Go through code to see if I can refactor things.~~

~~TODO: test for event_with_properties~~

## How did you test this code?

Unit tests

### CODE REVIEW

File: ee/clickhouse/queries/funnels/funnel_correlation.py [hunk 1]
```diff
@@ -380,16 +379,24 @@ def _get_events_join_query(self) -> str:
...
+        """
 
-            -- NOTE: I would love to right join here, so we count get total
-            -- success/failure numbers in one pass, but this causes out of memory
-            -- error mentioning issues with right filling. I'm sure there's a way
-            -- to do it but lifes too short.
+        aggregation_group_join = f"""
             JOIN funnel_people AS person
-                ON pdi.person_id = person.person_id
+                ON person.person_id = events.$group_{self._filter.aggregation_group_type_index}
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/funnel_correlation.py:396]: kinda annoying, `aggregation_target` is called `person_id` in funnel people CTEs. Updating that implies updating things in Paths as well, so didn't want to do this now. Perhaps once we start supporting "persons" modal.

File: ee/clickhouse/queries/funnels/funnel_correlation.py [hunk 2]
```diff
@@ -415,16 +422,41 @@ def _get_events_join_query(self) -> str:
...
+            ClickhousePersonQuery.PERSON_PROPERTIES_ALIAS
+            if self._filter.aggregation_group_type_index is None
+            else group_properties_field
+        )
+
         if "$all" in cast(list, self._filter.correlation_property_names):
             return (
                 f"""
-            arrayMap(x -> x.1, JSONExtractKeysAndValuesRaw({ClickhousePersonQuery.PERSON_PROPERTIES_ALIAS})) as person_prop_keys,
+            arrayMap(x -> x.1, JSONExtractKeysAndValuesRaw({aggregation_properties_alias})) as person_prop_keys,
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/funnel_correlation.py:455]: confirm I tested this branch for properties correlation

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+     OFFSET 0 SETTINGS allow_experimental_window_functions = 1),
+       toDateTime('2020-01-14 23:59:59') AS date_to,
+       toDateTime('2020-01-01 00:00:00') AS date_from,
+       2 AS target_step,
+       ['user signed up', 'paid'] as funnel_step_names
+  SELECT event.event AS name,
+         countDistinctIf(person.person_id, person.steps = target_step) AS success_count,
+         countDistinctIf(person.person_id, person.steps <> target_step) AS failure_count
+  FROM events AS event
+  JOIN funnel_people AS person ON person.person_id = events.$group_0
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:88]: Can I make this clearer here? despite the `funnel_people.person_id` constraint?

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:201]: No extra filtering possible on correlation by events. Yes. So no test with extra filters here possible.

(But event_with_property will have more interesting case here - TODO)

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+                             FROM person_distinct_id
+                             WHERE team_id = 2
+                             GROUP BY person_id,
+                                      distinct_id,
+                                      team_id
+                             HAVING max(is_deleted) = 0)
+                          GROUP BY distinct_id) AS pdi ON events.distinct_id = pdi.distinct_id
+                       INNER JOIN
+                         (SELECT group_key,
+                                 argMax(group_properties, _timestamp) AS group_properties_0
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:286]: Do a random test somewhere with a different group type index.

Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:286]: This is annoying? Group_properties were queried in the funnel, but they're only needed in the correlation query

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+            funnel_people.steps as steps,
+            arrayJoin(arrayZip(['industry'], [trim(BOTH '"' FROM JSONExtractRaw(groups_0.group_properties_0, 'industry'))])) as prop
+     FROM funnel_people
+     INNER JOIN
+       (SELECT group_key,
+               argMax(group_properties, _timestamp) AS group_properties_0
+        FROM groups
+        WHERE team_id = 2
+          AND group_type_index = 0
+        GROUP BY group_key) groups_0 ON funnel_people.person_id == groups_0.group_key) person_with_props
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:319]: don't call this person_with_props

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr (hunk 5 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:286]: that is, remove the funnel props from the CTE generator for funnels

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+  GROUP BY prop.1,
+               prop.2
+  HAVING prop.1 NOT IN []
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:435]: check??

File: ee/clickhouse/queries/funnels/test/test_funnel_correlation.py [hunk 8]
```diff
@@ -144,6 +146,160 @@ def test_basic_funnel_correlation_with_events(self):
         self.assertEqual(len(self._get_people_for_event(filter, "negatively_related", success=False)), 5)
         self.assertEqual(len(self._get_people_for_event(filter, "negatively_related")), 0)
 
+    def _create_groups(self):
```
Comment (neilkakkar) [inline ee/clickhouse/queries/funnels/test/test_funnel_correlation.py:149]: rm, not useful

File: ee/clickhouse/queries/groups_join_query.py [hunk 9]
```diff
@@ -29,10 +29,15 @@ def __init__(
         self._team_id = team_id
         self._column_optimizer = column_optimizer or ColumnOptimizer(self._filter, self._team_id)
 
-    def get_join_query(self) -> Tuple[str, Dict]:
+    def get_join_query(self, group_join_keys: Optional[List[str]] = None) -> Tuple[str, Dict]:
```
Comment (neilkakkar) [inline ee/clickhouse/queries/groups_join_query.py:32]: add tests for this

File: ee/clickhouse/queries/column_optimizer.py [hunk 10]
```diff
@@ -113,8 +113,13 @@ def properties_used_in_filter(self) -> Counter[PropertyIdentifier]:
                 counter += get_action_tables_and_properties(entity.get_action())
 
         if self.filter.correlation_type == FunnelCorrelationType.PROPERTIES and self.filter.correlation_property_names:
-            for prop_value in self.filter.correlation_property_names:
-                counter[(prop_value, "person", None)] += 1
+
+            if self.filter.aggregation_group_type_index is not None:
```
Comment (neilkakkar) [inline ee/clickhouse/queries/column_optimizer.py:117]: Add tests!

File: ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr (hunk 3 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/test/__snapshots__/test_funnel_correlation.ambr:88]: Hmm, makes things more confusing right now, as means updating table name else where, while keeping the `person_id` variable. So, opting not to do this right now.

File: ee/clickhouse/queries/groups_join_query.py [hunk 11]
```diff
@@ -29,10 +27,16 @@ def __init__(
...
         join_queries, params = [], {}
 
-        for group_type_index in self._column_optimizer.group_types_to_query:
+        if group_join_keys:
+            if len(group_join_keys) != len(self._column_optimizer.group_types_to_query):
+                raise ValidationError("When specifying group_join_keys, number should be equal to groups to be queried")
```
Comment (macobo) [inline ee/clickhouse/queries/groups_join_query.py:35]: This error is not gonna say much to the user. Is this an internal error? Should we raise an `ValueError` instead?

File: ee/clickhouse/queries/funnels/funnel_correlation.py [hunk 12]
```diff
@@ -415,16 +431,41 @@ def _get_events_join_query(self) -> str:
...
+                f"""
+                JOIN ({person_query}) person
+                    ON person.id = funnel_people.person_id
+            """,
+                person_query_params,
+            )
+        else:
+            return GroupsJoinQuery(
+                self._filter, self._team.pk, ColumnOptimizer(self._filter, self._team.pk)
+            ).get_join_query(["funnel_people.person_id"])
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation.py:450]: Proposal: Let's move this argument to GroupsJoinQuery constructor and add excessive documentation on what/why this exists.

Constructor since that's more in-line with how PersonQuery works.

File: ee/clickhouse/queries/groups_join_query.py [hunk 13]
```diff
@@ -29,10 +27,16 @@ def __init__(
         self._team_id = team_id
         self._column_optimizer = column_optimizer or ColumnOptimizer(self._filter, self._team_id)
 
-    def get_join_query(self) -> Tuple[str, Dict]:
+    def get_join_query(self, group_join_keys: Optional[List[str]] = None) -> Tuple[str, Dict]:
         join_queries, params = [], {}
 
-        for group_type_index in self._column_optimizer.group_types_to_query:
+        if group_join_keys:
```
Comment (macobo) [inline ee/clickhouse/queries/groups_join_query.py:33]: It's kind of iffy how the size of this needs to match `group_types_to_query` exactly - it creates a very tight coupling of things that shouldn't really be coupled (queries shouldn't know internals of column_optimizer).

Why is this needed? Can we decouple it somehow?

Looking at the callsite there'll only be a single argument passed so why not `str` only?

File: ee/clickhouse/queries/funnels/funnel_correlation.py [hunk 14]
```diff
@@ -415,16 +431,41 @@ def _get_events_join_query(self) -> str:
...
+            return (
+                f"""
+                JOIN ({person_query}) person
+                    ON person.id = funnel_people.person_id
+            """,
+                person_query_params,
+            )
+        else:
+            return GroupsJoinQuery(
+                self._filter, self._team.pk, ColumnOptimizer(self._filter, self._team.pk)
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation.py:449]: Nit: It's worth pulling this out into the constructor!

File: ee/clickhouse/queries/groups_join_query.py (hunk 13 above)
Comment (macobo) [reply ee/clickhouse/queries/groups_join_query.py:33]: My propsal is here: https://github.com/PostHog/posthog/pull/7056/files#r748075864

File: ee/clickhouse/queries/funnels/funnel_correlation.py (hunk 1 above)
Comment (macobo) [reply ee/clickhouse/queries/funnels/funnel_correlation.py:405]: Add a `:KLUDGE:` comment here?

File: ee/clickhouse/queries/funnels/funnel_correlation.py [hunk 15]
```diff
@@ -380,16 +388,24 @@ def _get_events_join_query(self) -> str:
...
-            -- success/failure numbers in one pass, but this causes out of memory
-            -- error mentioning issues with right filling. I'm sure there's a way
-            -- to do it but lifes too short.
+        aggregation_group_join = f"""
             JOIN funnel_people AS person
-                ON pdi.person_id = person.person_id
+                ON person.person_id = events.$group_{self._filter.aggregation_group_type_index}
+        """
+        return f"""
+            {aggregation_group_join if self._filter.aggregation_group_type_index is not None else aggregation_person_join}
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation.py:408]: Nit, Suggestion: Pull out method for this? The variable assignments above are slightly complex and deserves a name.

File: ee/clickhouse/queries/funnels/funnel_correlation.py (hunk 12 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/funnel_correlation.py:450]: There's only a single argument in this case, because we always expect to only have one group.

It made sense to me to keep an array with multiple possible options, since the join query can join multiple groups, and it wouldn't make sense to use the same alias to join on every group.

File: ee/clickhouse/queries/funnels/funnel_correlation.py (hunk 14 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/funnel_correlation.py:449]: and same for `ClickhousePersonQuery` ?

Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/funnel_correlation.py:449]: Ah, you mean the `GroupsJoinQuery` construction?

Comment (macobo) [reply ee/clickhouse/queries/funnels/funnel_correlation.py:449]: I meant ColumnOptimizer construction

Comment (neilkakkar) [reply ee/clickhouse/queries/funnels/funnel_correlation.py:449]: gotcha, yes, that makes more sense

File: ee/clickhouse/queries/funnels/funnel_correlation.py (hunk 12 above)
Comment (macobo) [reply ee/clickhouse/queries/funnels/funnel_correlation.py:450]: The array logic however doesn't really work - again, it's way too coupled, etc.

Rather than try to future-proof it I'd go with the const approach for now. Then later as the need arises we can e.g. refactor it to be a lambda or something like that easily.

File: ee/clickhouse/queries/groups_join_query.py [hunk 16]
```diff
@@ -24,16 +22,19 @@ def __init__(
...
         self._team_id = team_id
         self._column_optimizer = column_optimizer or ColumnOptimizer(self._filter, self._team_id)
+        self._join_key = join_key
 
     def get_join_query(self) -> Tuple[str, Dict]:
         join_queries, params = [], {}
 
         for group_type_index in self._column_optimizer.group_types_to_query:
             var = f"group_index_{group_type_index}"
+            group_join_key = self._join_key or f"$group_{group_type_index}"
```
Comment (macobo) [inline ee/clickhouse/queries/groups_join_query.py:37]: nit: you could also do this in the constructor.

**Review by macobo — APPROVED** [review]
I am happy with this :muscle:

File: ee/clickhouse/queries/groups_join_query.py (hunk 16 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/groups_join_query.py:37]: how would I get the index? O.o - or do you mean creating a lambda that takes an index?

Comment (macobo) [reply ee/clickhouse/queries/groups_join_query.py:37]: Nevermind :facepalm:
